File: frontend/src/features/applications/types/application.ts
export type ApplicationStatus =
  | "applied"
  | "phone_screen"
  | "onsite"
  | "offer"
  | "rejected";

export interface ApplicationItem {
  id: number;

  user_id: number;

  company_name: string;

  job_title: string;

  status: ApplicationStatus;

  applied_date: string;

  phone_screen_date?: string | null;

  onsite_date?: string | null;

  offer_date?: string | null;

  rejected_date?: string | null;

  follow_up_date?: string | null;

  notes: string;

  created_at: string;

  updated_at: string;
}

export interface ApplicationCreatePayload {
  company_name: string;

  job_title: string;

  status: ApplicationStatus;

  applied_date: string;

  phone_screen_date?: string | null;

  onsite_date?: string | null;

  offer_date?: string | null;

  rejected_date?: string | null;

  follow_up_date?: string | null;

  notes: string;
}

export type ApplicationUpdatePayload =
  Partial<ApplicationCreatePayload>;

export interface ApplicationStats {
  active: number;

  response_rate: number;

  offers: number;

  follow_ups_due: number;
}

export interface ApplicationListResponse {
  applications: ApplicationItem[];

  stats: ApplicationStats;
}

File: backend/app/schemas/applications/application.py
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel


ApplicationStatus = Literal[
    "applied",
    "phone_screen",
    "onsite",
    "offer",
    "rejected",
]


class ApplicationBase(BaseModel):
    company_name: str
    job_title: str
    status: ApplicationStatus = "applied"

    applied_date: date
    phone_screen_date: Optional[date] = None
    onsite_date: Optional[date] = None
    offer_date: Optional[date] = None
    rejected_date: Optional[date] = None

    follow_up_date: Optional[date] = None

    notes: str = ""


class ApplicationCreate(ApplicationBase):
    pass


class ApplicationUpdate(BaseModel):
    company_name: Optional[str] = None
    job_title: Optional[str] = None
    status: Optional[ApplicationStatus] = None

    applied_date: Optional[date] = None
    phone_screen_date: Optional[date] = None
    onsite_date: Optional[date] = None
    offer_date: Optional[date] = None
    rejected_date: Optional[date] = None

    follow_up_date: Optional[date] = None

    notes: Optional[str] = None


class ApplicationResponse(ApplicationBase):
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ApplicationStats(BaseModel):
    active: int
    response_rate: int
    offers: int
    follow_ups_due: int


class ApplicationListResponse(BaseModel):
    applications: list[ApplicationResponse]
    stats: ApplicationStats
